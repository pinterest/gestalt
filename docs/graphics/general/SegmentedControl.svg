<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><rect width="218" height="30" x="47" y="75" fill="#005C62" rx="15"/><rect width="74" height="24" x="50" y="78" fill="#fff" rx="12"/><rect width="48" height="8" x="138" y="86" fill="#CCF6EE" fill-opacity=".5" rx="4"/><rect width="48" height="8" x="202" y="86" fill="#CCF6EE" fill-opacity=".5" rx="4"/><rect width="218" height="30" x="47" y="75" stroke="#005C62" stroke-width="2" rx="15"/><rect width="58" height="8" x="58" y="86" fill="#009990" fill-opacity=".5" rx="4"/></svg>