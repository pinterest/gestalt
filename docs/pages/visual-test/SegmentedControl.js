// @flow strict
import { type Node as ReactNode, useState } from 'react';
import { Box, Flex, Heading, SegmentedControl } from 'gestalt';

export default function SegmentedControlSnapshot(): ReactNode {
  const [item1Index, setItem1Index] = useState(0);
  const [item2Index, setItem2Index] = useState(0);
  const [item3Index, setItem3Index] = useState(0);
  const items = ['Small', 'Medium', 'Large'];

  return (
    <Box color="default" display="inlineBlock" padding={1}>
      <Box padding={8}>
        <Flex direction="column" gap={{ column: 6, row: 0 }}>
          <Flex direction="column" gap={{ column: 2, row: 0 }}>
            <Heading size="300">sm</Heading>
            <SegmentedControl
              items={items}
              onChange={({ activeIndex }) => {
                setItem1Index(activeIndex);
              }}
              selectedItemIndex={item1Index}
              size="sm"
            />
          </Flex>

          <Flex direction="column" gap={{ column: 2, row: 0 }}>
            <Heading size="300">md</Heading>
            <SegmentedControl
              items={items}
              onChange={({ activeIndex }) => {
                setItem2Index(activeIndex);
              }}
              responsive
              selectedItemIndex={item2Index}
              size="md"
            />
          </Flex>
          <Flex direction="column" gap={{ column: 2, row: 0 }}>
            <Heading size="300">lg</Heading>
            <SegmentedControl
              items={items}
              onChange={({ activeIndex }) => {
                setItem3Index(activeIndex);
              }}
              responsive
              selectedItemIndex={item3Index}
              size="lg"
            />
          </Flex>
        </Flex>
      </Box>
    </Box>
  );
}
