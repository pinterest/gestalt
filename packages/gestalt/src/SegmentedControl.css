.SegmentedControl {
  composes: borderBox from "./Layout.css";
  composes: lightGrayBg from "./Colors.css";
  composes: flex from "./Layout.css";
  composes: justifyBetween from "./Layout.css";
  border-radius: 4px;
  padding: 2px;
  user-select: none;
}

.md {
  composes: medium from "./Layout.css";
}

.lg {
  composes: large from "./Layout.css";
}

.item {
  composes: accessibilityOutline from "./Focus.css";
  composes: flexGrow from "./Layout.css";
  composes: m0 from "./Whitespace.css";
  composes: noBorder from "./Borders.css";
  composes: pointer from "./Cursor.css";
  border-radius: 3px;
  flex-basis: 0;
  flex-shrink: 1;
  padding: 4px 14px;
}

.item:focus {
  position: relative;
}

.itemIsNotSelected {
  background: transparent;
}

.itemIsSelected {
  composes: whiteBg from "./Colors.css";
}
