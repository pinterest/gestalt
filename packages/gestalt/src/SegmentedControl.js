// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Box from './Box.js';
import Text from './Text.js';
import styles from './SegmentedControl.css';

type Props = {|
  items: Array<React.Node>,
  onChange: ({ event: SyntheticMouseEvent<>, activeIndex: number }) => void,
  responsive?: boolean,
  selectedItemIndex: number,
  size?: 'md' | 'lg',
|};

export default function SegmentedControl(props: Props) {
  const { items, onChange, responsive, selectedItemIndex, size = 'md' } = props;
  const buttonWidth = responsive
    ? undefined
    : `${Math.floor(100 / Math.max(1, items.length))}%`;
  return (
    <div
      className={classnames(styles.SegmentedControl, {
        [styles.md]: size === 'md',
        [styles.lg]: size === 'lg',
      })}
      role="tablist"
    >
      {items.map((item, i) => {
        const isSelected = i === selectedItemIndex;
        const cs = classnames(styles.item, {
          [styles.itemIsNotSelected]: !isSelected,
          [styles.itemIsSelected]: isSelected,
        });
        return (
          <button
            aria-selected={isSelected}
            className={cs}
            key={i}
            onClick={e => onChange({ event: e, activeIndex: i })}
            role="tab"
            type="button"
            style={{ width: buttonWidth }}
          >
            {typeof item === 'string' ? (
              <Text
                bold
                color={isSelected ? 'darkGray' : 'gray'}
                align="center"
                size={size}
              >
                {item}
              </Text>
            ) : (
              <Box display="flex" justifyContent="center">
                {item}
              </Box>
            )}
          </button>
        );
      })}
    </div>
  );
}

SegmentedControl.propTypes = {
  items: PropTypes.arrayOf(PropTypes.node).isRequired,
  onChange: PropTypes.func.isRequired,
  responsive: PropTypes.bool,
  selectedItemIndex: PropTypes.number.isRequired,
  size: PropTypes.oneOf(['md', 'lg']),
};
