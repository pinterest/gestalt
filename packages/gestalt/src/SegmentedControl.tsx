import { cloneElement, ReactNode } from 'react';
import classnames from 'classnames';
import Box from './Box';
import focusStyles from './Focus.css';
import layout from './Layout.css';
import styles from './SegmentedControl.css';
import Text from './Text';
import useFocusVisible from './useFocusVisible';

type SizeType = 'sm' | 'md' | 'lg';

type OnChange = (arg1: { event: React.MouseEvent<HTMLButtonElement>; activeIndex: number }) => void;

type Props = {
  /**
   * Items for selection. Though typically strings, React.Node is accepted to allow for Icons or other custom UI.
   */
  items: ReadonlyArray<ReactNode>;
  /**
   * Callback triggered when the user selects an item.
   */
  onChange: OnChange;
  /**
   * By default, items have equal widths. If this prop is true, the width of an item is based on its content. See the [responsive example](https://gestalt.pinterest.systems/web/segmentedcontrol#Example:-Responsive) for more details.
   */
  responsive?: boolean;
  /**
   * Index of element in `items` that is currently selected.
   */
  selectedItemIndex: number;
  /**
   * Size of the Segmented Control.
   */
  size?: SizeType;
};

const getDensityStyles = (s: SizeType) => {
  switch (s) {
    case 'sm':
      return { fontSize: '200', iconSize: 16 };
    case 'lg':
      return { fontSize: '300', iconSize: 24 };
    case 'md':
    default:
      return { fontSize: '300', iconSize: 20 };
  }
};
const applyDensityStyle = (s: SizeType) => styles[`${s}`];

// layout.css contains a mapping of size to min-height
const applyMinHeight = (s: SizeType) => {
  const lookup = { 'sm': 'small', 'md': 'medium', 'lg': 'large' } as const;
  return layout[`${lookup[s]}`];
};

function SegmentedControlItem({
  index,
  item,
  isSelected,
  onChange,
  size = 'md',
  width,
}: {
  index: number;
  item: ReactNode;
  isSelected: boolean;
  onChange: OnChange;
  width: string | null | undefined;
  size: SizeType;
}) {
  const { isFocusVisible } = useFocusVisible();
  const cs = classnames(
    styles.item,
    focusStyles.hideOutline,
    {
      [styles.itemIsNotSelected]: !isSelected,
      [styles.itemIsSelected]: isSelected,
      [focusStyles.accessibilityOutline]: isFocusVisible,
    },
    applyDensityStyle(size),
  );

  const { fontSize, iconSize } = getDensityStyles(size);

  return (
    <button
      aria-selected={isSelected}
      className={cs}
      onClick={(event) => onChange({ event, activeIndex: index })}
      role="tab"
      // @ts-expect-error - TS2322 - Type 'string | null | undefined' is not assignable to type 'Width<string | number> | undefined'.
      style={{ width }}
      type="button"
    >
      {typeof item === 'string' ? (
        // @ts-expect-error - TS2322 - Type 'string' is not assignable to type 'Size | undefined'.
        <Text align="center" color="default" size={fontSize} weight="bold">
          {item}
        </Text>
      ) : (
        <Box display="flex" justifyContent="center">
          {/* @ts-expect-error - TS2339 - Property 'type' does not exist on type 'ReactElement<any, string | JSXElementConstructor<any>> | Iterable<ReactNode> | ReactPortal'. | TS2339 - Property 'type' does not exist on type 'ReactElement<any, string | JSXElementConstructor<any>> | Iterable<ReactNode> | ReactPortal'. */}
          {typeof item === 'object' && item && item.type && item.type.displayName === 'Icon'
            ? // @ts-expect-error - TS2769 - No overload matches this call.
              cloneElement(item, { size: iconSize })
            : item}
        </Box>
      )}
    </button>
  );
}

/**
 * [SegmentedControl](https://gestalt.pinterest.systems/web/segmentedcontrol)  may be used to group multiple selections. The controls display the current state and related state.
 *
 * Create layout to convey clear sense of information hierarchy. When a control is engaged, information below the control should also be updated.
 *
 * ![SegmentedControl light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/SegmentedControl.spec.mjs-snapshots/SegmentedControl-chromium-darwin.png)
 * ![SegmentedControl dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/SegmentedControl-dark.spec.mjs-snapshots/SegmentedControl-dark-chromium-darwin.png)
 *
 */
export default function SegmentedControl({
  items,
  onChange,
  responsive,
  selectedItemIndex,
  size = 'md',
}: Props) {
  const buttonWidth = responsive ? undefined : `${Math.floor(100 / Math.max(1, items.length))}%`;
  return (
    <div
      className={classnames(styles.SegmentedControl, applyMinHeight(size), applyDensityStyle(size))}
      role="tablist"
    >
      {items.map((item, i) => (
        <SegmentedControlItem
          // eslint-disable-next-line react/no-array-index-key
          key={i}
          index={i}
          isSelected={i === selectedItemIndex}
          item={item}
          onChange={onChange}
          size={size}
          width={buttonWidth}
        />
      ))}
    </div>
  );
}

SegmentedControl.displayName = 'SegmentedControl';
