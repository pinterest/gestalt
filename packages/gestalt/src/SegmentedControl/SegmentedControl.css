.SegmentedControl {
  composes: lightGrayBg from "../Colors.css";
  composes: flex from "../Layout.css";
  composes: justifyBetween from "../Layout.css";
  border-radius: 4px;
  padding: 2px;
  user-select: none;
}

.item {
  composes: accessibilityOutline from "../Focus.css";
  composes: flexGrow from "../Layout.css";
  composes: m0 from "../Whitespace.css";
  composes: noBorder from "../Borders.css";
  composes: pointer from "../Cursor.css";
  border-radius: 3px;
  flex-basis: 0;
  flex-shrink: 1;
}

.md {
  padding: 9px 14px;
}

.lg {
  padding: 12px 14px;
}

.item:focus {
  position: relative;
}

.itemIsNotSelected {
  background: transparent;
}

.itemIsSelected {
  composes: whiteBg from "../Colors.css";
}
