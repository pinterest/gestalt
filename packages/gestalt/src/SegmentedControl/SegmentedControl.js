// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Text from '../Text/Text';
import styles from './SegmentedControl.css';

type Props = {|
  items: Array<string>,
  onChange: ({ event: SyntheticMouseEvent<>, activeIndex: number }) => void,
  selectedItemIndex: number,
  size?: 'md' | 'lg',
|};

export default function SegmentedControl(props: Props) {
  const { items, onChange, selectedItemIndex, size = 'md' } = props;
  return (
    <div className={styles.SegmentedControl} role="tablist">
      {items.map((item, i) => {
        const isSelected = i === selectedItemIndex;
        const cs = classnames(styles.item, {
          [styles.md]: size === 'md',
          [styles.lg]: size === 'lg',
          [styles.itemIsNotSelected]: !isSelected,
          [styles.itemIsSelected]: isSelected,
        });
        return (
          <button
            aria-selected={isSelected}
            className={cs}
            key={i}
            onClick={e => onChange({ event: e, activeIndex: i })}
            role="tab"
          >
            <Text
              bold
              color={isSelected ? 'darkGray' : 'gray'}
              align="center"
              size={size}
            >
              {item}
            </Text>
          </button>
        );
      })}
    </div>
  );
}

SegmentedControl.propTypes = {
  items: PropTypes.arrayOf(PropTypes.node).isRequired,
  onChange: PropTypes.func.isRequired,
  selectedItemIndex: PropTypes.number.isRequired,
};
