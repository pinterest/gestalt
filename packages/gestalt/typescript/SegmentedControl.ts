import type { Node } from "react";
import classnames from "classnames";
import type { AbstractEventHandler } from "./AbstractEventHandler";
import "./AbstractEventHandler";
import Box from "./Box";
import focusStyles from "./Focus.css";
import layout from "./Layout.css";
import styles from "./SegmentedControl.css";
import Text from "./Text";
import useFocusVisible from "./useFocusVisible";
type OnChange = AbstractEventHandler<
  React.MouseEvent<HTMLButtonElement>,
  {
    activeIndex: number;
  }
>;
type Props = {
  items: ReadonlyArray<Node>;
  onChange: OnChange;
  responsive?: boolean;
  selectedItemIndex: number;
  size?: "md" | "lg";
};

function SegmentedControlItem({
  index,
  item,
  isSelected,
  onChange,
  size,
  width,
}: {
  index: number;
  item: Node;
  isSelected: boolean;
  onChange: OnChange;
  size?: "md" | "lg";
  width: string | null | undefined;
}) {
  const { isFocusVisible } = useFocusVisible();
  const cs = classnames(styles.item, focusStyles.hideOutline, {
    [styles.itemIsNotSelected]: !isSelected,
    [styles.itemIsSelected]: isSelected,
    [focusStyles.accessibilityOutline]: isFocusVisible,
  });
  return (
    <button
      aria-selected={isSelected}
      className={cs}
      onClick={(event) =>
        onChange({
          event,
          activeIndex: index,
        })
      }
      role="tab"
      type="button"
      style={{
        width,
      }}
    >
      {typeof item === "string" ? (
        <Text color="darkGray" align="center" size={size} weight="bold">
          {item}
        </Text>
      ) : (
        <Box display="flex" justifyContent="center">
          {item}
        </Box>
      )}
    </button>
  );
}
/**
 * https://gestalt.pinterest.systems/SegmentedControl
 */

export default function SegmentedControl({
  items,
  onChange,
  responsive,
  selectedItemIndex,
  size = "md",
}: Props): Node {
  const buttonWidth = responsive
    ? undefined
    : `${Math.floor(100 / Math.max(1, items.length))}%`;
  return (
    <div
      className={classnames(
        styles.SegmentedControl,
        size === "md" ? layout.medium : layout.large
      )}
      role="tablist"
    >
      {items.map((item, i) => (
        <SegmentedControlItem
          key={i}
          index={i}
          item={item}
          isSelected={i === selectedItemIndex}
          onChange={onChange}
          size={size}
          width={buttonWidth}
        />
      ))}
    </div>
  );
}