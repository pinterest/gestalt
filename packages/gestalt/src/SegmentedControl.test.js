// @flow
import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import SegmentedControl from './SegmentedControl.js';

test('SegmentedControl renders', () => {
  const tree = create(
    <SegmentedControl
      items={['News', 'You', 'Messages', <div key="dummy" />]}
      selectedItemIndex={0}
      onChange={() => {}}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('SegmentedControl with responsive widths renders', () => {
  const tree = create(
    <SegmentedControl
      responsive
      items={['Short', 'Really really really long title']}
      selectedItemIndex={0}
      onChange={() => {}}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('SegmentedControl handles click', () => {
  const mockOnChange = jest.fn();
  const wrapper = shallow(
    <SegmentedControl
      items={['Item']}
      selectedItemIndex={0}
      onChange={mockOnChange}
    />
  );

  const mockEvent = {};
  wrapper.find('button').simulate('click', mockEvent);

  expect(mockOnChange).toBeCalled();
  expect(mockOnChange).toBeCalledWith({ event: mockEvent, activeIndex: 0 });
});
