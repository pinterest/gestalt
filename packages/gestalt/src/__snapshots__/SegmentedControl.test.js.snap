// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedControl renders 1`] = `
<div
  className="SegmentedControl md"
  role="tablist"
>
  <button
    aria-selected={true}
    className="item itemIsSelected"
    onClick={[Function]}
    role="tab"
    style={
      Object {
        "width": "25%",
      }
    }
    type="button"
  >
    <div
      className="Text fontSize3 darkGray alignCenter breakWord fontStyleNormal fontWeightBold"
    >
      News
    </div>
  </button>
  <button
    aria-selected={false}
    className="item itemIsNotSelected"
    onClick={[Function]}
    role="tab"
    style={
      Object {
        "width": "25%",
      }
    }
    type="button"
  >
    <div
      className="Text fontSize3 gray alignCenter breakWord fontStyleNormal fontWeightBold"
    >
      You
    </div>
  </button>
  <button
    aria-selected={false}
    className="item itemIsNotSelected"
    onClick={[Function]}
    role="tab"
    style={
      Object {
        "width": "25%",
      }
    }
    type="button"
  >
    <div
      className="Text fontSize3 gray alignCenter breakWord fontStyleNormal fontWeightBold"
    >
      Messages
    </div>
  </button>
  <button
    aria-selected={false}
    className="item itemIsNotSelected"
    onClick={[Function]}
    role="tab"
    style={
      Object {
        "width": "25%",
      }
    }
    type="button"
  >
    <div
      className="box justifyCenter xsDisplayFlex"
    >
      <div />
    </div>
  </button>
</div>
`;

exports[`SegmentedControl with responsive widths renders 1`] = `
<div
  className="SegmentedControl md"
  role="tablist"
>
  <button
    aria-selected={true}
    className="item itemIsSelected"
    onClick={[Function]}
    role="tab"
    style={
      Object {
        "width": undefined,
      }
    }
    type="button"
  >
    <div
      className="Text fontSize3 darkGray alignCenter breakWord fontStyleNormal fontWeightBold"
    >
      Short
    </div>
  </button>
  <button
    aria-selected={false}
    className="item itemIsNotSelected"
    onClick={[Function]}
    role="tab"
    style={
      Object {
        "width": undefined,
      }
    }
    type="button"
  >
    <div
      className="Text fontSize3 gray alignCenter breakWord fontStyleNormal fontWeightBold"
    >
      Really really really long title
    </div>
  </button>
</div>
`;
